sway/commands/bar.c

```diff
@@ -36,54 +34,49 @@ static struct cmd_handler bar_handlers[] = {
 
 // Must be in alphabetical order for bsearch
 static struct cmd_handler bar_config_handlers[] = {
-	{ "hidden_state", bar_cmd_hidden_state },
-	{ "mode", bar_cmd_mode }
+	{ "id", bar_cmd_id },
+	{ "swaybar_command", bar_cmd_swaybar_command },
 };
 
+// Determines whether the subcommand is valid in any bar handler struct
+static bool is_subcommand(char *name) {
+	return find_handler(name, bar_handlers, sizeof(bar_handlers)) ||
+		find_handler(name, bar_config_handlers, sizeof(bar_config_handlers));
+}
+
 struct cmd_results *cmd_bar(int argc, char **argv) {
 	struct cmd_results *error = NULL;
-	if ((error = checkarg(argc, "bar", EXPECTED_AT_LEAST, 1))) {
+	if ((error = checkarg(argc, "bar", EXPECTED_AT_LEAST, 2))) {
 		return error;
 	}
 
-	if (find_handler(argv[0], bar_config_handlers,
-				sizeof(bar_config_handlers))) {
-		if (config->reading) {
-			return config_subcommand(argv, argc, bar_config_handlers,
-					sizeof(bar_config_handlers));
-		}
-		return cmd_results_new(CMD_FAILURE, "bar",
-				"Can only be used in config file.");
-	}
-
-	if (argc > 1) {
-		struct bar_config *bar = NULL;
-		if (!find_handler(argv[0], bar_handlers, sizeof(bar_handlers))
-				&& find_handler(argv[1], bar_handlers, sizeof(bar_handlers))) {
-			for (int i = 0; i < config->bars->length; ++i) {
-				struct bar_config *item = config->bars->items[i];
-				if (strcmp(item->id, argv[0]) == 0) {
-					wlr_log(WLR_DEBUG, "Selecting bar: %s", argv[0]);
-					bar = item;
-					break;
-				}
+	bool spawn = false;
+	struct bar_config *bar = NULL;
+	if (strcmp(argv[0], "id") != 0 && is_subcommand(argv[1])) {
+		for (int i = 0; i < config->bars->length; ++i) {
+			struct bar_config *item = config->bars->items[i];
+			if (strcmp(item->id, argv[0]) == 0) {
+				wlr_log(WLR_DEBUG, "Selecting bar: %s", argv[0]);
+				bar = item;
+				break;
 			}
+		}
+		if (!bar) {
+			spawn = !config->reading;
```

Comment:
Would it be possible to change spawning behaviour so it only needs to allocate once?

I'm thinking something like using a variable like `spawn_with_id`.

I'm also confused why this loop is required later:

https://github.com/swaywm/sway/blob/782a835175b1fecb427fbbafef4e7518af95329f/sway/commands/bar.c#L95

since wouldn't the bar always be the last one?

Anyway, I think this may be outside of the scope of this PR, and I may open my own one. It otherwise looks good to me, and if you don't feel like dealing with this I'm happy to accept it.

Reply:
> Would it be possible to change spawning behaviour so it only needs to allocate once? I'm thinking something like using a variable like spawn_with_id.

I'm not sure I understand what you are asking. That's line is inside the creation of a bar with a custom id. `spawn` would not be set to `true` if the bar already exists.

> I'm also confused why this loop is required later [..] since wouldn't the bar always be the last one?

I'm actually not sure. I'm guessing its an artifact from an older version of the codebase. I'll remove the loop

Reply:
The loop has been removed. I also fixed up a few of the commits to group them better since there were multiple small related changes.